src/util.c

```diff
@@ -709,6 +709,68 @@ get_unspecified_address(int ip_family)
     return (ip_family == AF_INET) ? "0.0.0.0" : "::";
 }
 
+struct netdef_address_iter {
+    guint ip4_index;
+    guint ip6_index;
+    guint address_options_index;
+    NetplanNetDefinition* netdef;
+};
+
+NETPLAN_INTERNAL struct netdef_address_iter*
+_netplan_new_netdef_address_iter(NetplanNetDefinition* netdef)
+{
+    struct netdef_address_iter* it = g_malloc0(sizeof(struct netdef_address_iter));
+    it->ip4_index = 0;
+    it->ip6_index = 0;
+    it->address_options_index = 0;
+    it->netdef = netdef;
+
+    return it;
+}
+
+/*
+ * The netdef address iterator produces NetplanAddressOptions
+ * for all the addresses stored in ip4_address, ip6_address and
+ * address_options (in this order).
+ *
+ * The user of the iterator is responsible for calling
```

Comment:
We could also store the list of `NetplanAddressOptions` in the `netdef_address_iter` struct and release them when the iterator is destroyed. But with a big number of IPs we would end up having a lot of memory allocated for the iterator lifetime, while with this simpler approach we will keep a single one and release it in the Python code.

Another approach is to convert `netdef.ip4_addresses` and `netdef.ip6_addresses` from char* to `NetplanAddressOptions*`. With that we would simply send the pointer to the caller and no extra memory would need to be allocated...